lib/std/priority_dequeue.zig

```diff
@@ -0,0 +1,879 @@
+            if (start.min_layer) {
+                doSiftUp(self, start.index, lessThan);
+            } else {
+                doSiftUp(self, start.index, greaterThan);
+            }
+        }
+
+        fn doSiftUp(self: *Self, start_index: usize, compare: fn (Self, T, T) bool) void {
+            var child_index = start_index;
+            while (child_index > 2) {
+                var grandparent_index = grandparentIndex(child_index);
+                const child = self.items[child_index];
+                const grandparent = self.items[grandparent_index];
+
+                // If the grandparent is already better, we have gone as far as we need to
+                if (!compare(self.*, child, grandparent)) break;
+
+                // Otherwise swap the item with it's grandparent
+                self.items[grandparent_index] = child;
+                self.items[child_index] = grandparent;
+                child_index = grandparent_index;
+            }
+        }
+
+        /// Look at the smallest element in the dequeue. Returns
+        /// `null` if empty.
+        pub fn peekMin(self: *Self) ?T {
+            return if (self.len > 0) self.items[0] else null;
+        }
+
+        /// Look at the largest element in the dequeue. Returns
+        /// `null` if empty.
+        pub fn peekMax(self: *Self) ?T {
+            if (self.len == 0) return null;
+            if (self.len == 1) return self.items[0];
+            if (self.len == 2) return self.items[1];
+            return self.bestItemAtIndices(1, 2, greaterThan).item;
+        }
+
+        fn maxIndex(self: Self) ?usize {
+            if (self.len == 0) return null;
+            if (self.len == 1) return 0;
+            if (self.len == 2) return 1;
+            return self.bestItemAtIndices(1, 2, greaterThan).index;
+        }
+
+        /// Pop the smallest element from the dequeue. Returns
+        /// `null` if empty.
+        pub fn removeMinOrNull(self: *Self) ?T {
+            return if (self.len > 0) self.removeMin() else null;
+        }
+
+        /// Remove and return the smallest element from the
+        /// dequeue.
+        pub fn removeMin(self: *Self) T {
+            return self.removeIndex(0);
+        }
+
+        /// Pop the largest element from the dequeue. Returns
+        /// `null` if empty.
+        pub fn removeMaxOrNull(self: *Self) ?T {
+            return if (self.len > 0) self.removeMax() else null;
+        }
+
+        /// Remove and return the largest element from the
+        /// dequeue.
+        pub fn removeMax(self: *Self) T {
+            return self.removeIndex(self.maxIndex().?);
+        }
+
+        /// Remove and return element at index. Indices are in the
+        /// same order as iterator, which is not necessarily priority
+        /// order.
+        pub fn removeIndex(self: *Self, index: usize) T {
+            const item = self.items[index];
+            const last = self.items[self.len - 1];
```

Comment:
Missing guard against `self.len==0`? There's also `mem.swap` to clean up this code a bit.

Reply:
Added the guard. Are you suggesting `mem.swap` in this function, or lower down to replace segments that look like this?

```
self.items[grandparent_index] = child;
self.items[child_index] = grandparent;
```

In this particular function I'm not sure it adds anything, as we only have to move one item. I can see the argument for replacing these other blocks though.

Reply:
That was more of a general reminder, whenever you need to swap two elements and cannot be arsed to write the usual boilerplate `mem.swap` comes in handy